File: frontend/subjects/src/services/subject.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Subject } from '../app/model/subjects';

@Injectable({ providedIn: 'root' })
export class SubjectService {
  private baseUrl = 'http://localhost:9090/api/subjects';
  constructor(private http: HttpClient) {}

  getAll(): Observable<Subject[]> {
    return this.http.get<Subject[]>(this.baseUrl);
  }

  create(subject: Subject): Observable<Subject> {
    return this.http.post<Subject>(this.baseUrl, subject);
  }

  update(id: string, subject: Subject): Observable<Subject> {
    return this.http.put<Subject>(`${this.baseUrl}/${id}`, subject);
  }

  delete(id: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }
}

File: backend/subjects/src/main/java/com/example/subjects/service/SubjectService.java
package com.example.subjects.service;

import com.example.subjects.model.Subject;
import com.example.subjects.model.dtos.SubjectDto;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public interface SubjectService {
    // Define service methods here
    List<SubjectDto> getAll();
    SubjectDto get(String id);
    SubjectDto create(SubjectDto subjectDto);
    SubjectDto update(String id, SubjectDto subjectDto);
    void delete(String id);
}
